feat: Extend the text based configuration to include feature flags and the SaturationDetector's configuration

shmuelk opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Eliminates the use of environment variables to enable the new Data Layer and configure the SaturationDetector.

The use of environment variables for these types of things is not scalable. It also much easer to maintain multiple configurations if needed in the form of simple YAML text.

**Which issue(s) this PR fixes**:
Fixes #1481

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
- Configuring the saturation detector via the environment variables `SD_QUEUE_DEPTH_THRESHOLD`, `SD_KV_CACHE_UTIL_THRESHOLD`, and `SD_METRICS_STALENESS_THRESHOLD` is now DEPRECATED. Instead it should be configured via the  `saturationDetector` section of the text based configuration. See the text based configuration guide for more details.
- Enabling the experimental Data Layer feature via the `ENABLE_EXPERIMENTAL_DATALAYER_V2` environment variable is now DEPRECATED. Instead it should be enabled in the `featureGates` section of the text based configuration. See the text based configuration guide for more details.
- Enabling the experimental Flow Control feature via the `ENABLE_EXPERIMENTAL_FLOW_CONTROL_LAYER` environment variable is now DEPRECATED. Instead it should be enabled in the `featureGates` section of the text based configuration. See the text based configuration guide for more details.
```

nirrozenbaum commented on apix/config/v1alpha1/endpointpickerconfig_types.go:
```diff
@@ -118,3 +130,68 @@ func (sp SchedulingPlugin) String() string {
 }
+
+// FeatureGates is a set of flags that enable various experimental features with the EPP
+type FeatureGates struct {
+	// +optional
+	// EnableDataLayer If present and true enables the experimental Datalayer APIs.
+	// The Datalayer APIswill be disabled if EnableDataLayer is not present in the
+	// configuration or its value is false.
+	EnableDataLayer bool `json:"enableDataLayer,omitempty"`
+
+	// +optional
+	// EnableFlowControl If present and true enables the experimental FlowControl feature.
+	// The FlowControl feature will be disabled if EnableFlowControl is not in the
+	// configuration or its value is false.
+	EnableFlowControl bool `json:"enableFlowControl,omitempty"`
+}
```
feature gates by definition are boolean to enable/disable a feature. 
can we just call it `DataLayer` and `FlowControl`?

nirrozenbaum commented on apix/config/v1alpha1/endpointpickerconfig_types.go:
```diff
@@ -39,13 +39,25 @@ type EndpointPickerConfig struct {
+	// +optional
+	// SaturationDetector when present specifies the configuration of the
+	// Saturation detector. If not present, default values are used.
+	SaturationDetector *SaturationDetector `json:"saturationDetector,omitempty"`
```
I think this is too specific and should be generalized. 
we've talked in the past about having a parameters section in the config API, such that multiple structs in the code can consume. for example, I might want to consume the metricsStalenessThreshold in more places.
This approach is not future proof and we might need to add more and more fields here instead of just generalizing.

nirrozenbaum commented on apix/config/v1alpha1/endpointpickerconfig_types.go:
```diff
@@ -118,3 +130,68 @@ func (sp SchedulingPlugin) String() string {
 }
+
+// FeatureGates is a set of flags that enable various experimental features with the EPP
+type FeatureGates struct {
+	// +optional
+	// EnableDataLayer If present and true enables the experimental Datalayer APIs.
+	// The Datalayer APIswill be disabled if EnableDataLayer is not present in the
+	// configuration or its value is false.
+	EnableDataLayer bool `json:"enableDataLayer,omitempty"`
+
+	// +optional
+	// EnableFlowControl If present and true enables the experimental FlowControl feature.
+	// The FlowControl feature will be disabled if EnableFlowControl is not in the
+	// configuration or its value is false.
+	EnableFlowControl bool `json:"enableFlowControl,omitempty"`
+}
```
can this be generalized to a `map[string]bool` instead of two boolean fields? what if tomorrow I have another feature? would we need to update the config api code on every new feature?
what if I'd like to use the same config API to enable a feature in llm-d that is not part of the GIE?

shmuelk replied:
The disadvantage of a map and environment variables for that matter is that it is difficult to validate. A silly typo causes all sorts of problems.

Validating the keys, will not solve your desire of experimental issues in llm-d

shmuelk replied:
Having said that, let me give this some more thought.

shmuelk commented on pkg/epp/config/loader/configloader.go:
```diff
@@ -116,6 +125,41 @@ func loadSchedulerConfig(configProfiles []configapi.SchedulingProfile, handle pl
+	return featureConfig
+}
+
+func loadSaturationDetectorConfig(sd *configapi.SaturationDetector) saturationdetector.Config {
```
There were previous reviews of the original text based configuration PR in which there was a desire to keep all of the defaults setting code in one place to make it easier to understand the overall configuration.

There was also a great desire to separate validation from setting of defaults.

I would assume the same is basically true for the validation of the configuration. At this time the validation code is inside configloader.go. It could easily be moved in its own source file with a set of functions each validating a different part of the configuration.

shmuelk replied:
I have moved the validation code into a separate file (validation.go) in the pkg/epp/config/loader directory

shmuelk commented:
>>  @shmuelk I have not reviewed this PR yet, but I want to comment on the release notes:

>> - The saturation detector is no longer configured via the environment variables `SD_QUEUE_DEPTH_THRESHOLD`,  `SD_KV_CACHE_UTIL_THRESHOLD`, and `SD_METRICS_STALENESS_THRESHOLD`. Instead it is configured via the  `saturationDetector` section of the text based configuration. See the text based configuration guide for more details.
>> - The experimental Data Layer feature is no longer enabled via the `ENABLE_EXPERIMENTAL_DATALAYER_V2` environment variable. It is instead enabled in the `featureGates` section of the text based configuration. See the text based configuration guide for more details.
>> We should deprecate environment variable-based config for a release cycle before removing it. During deprecation, a log message should be generated that the env var used is deprecated and env var-based config takes precedence.

>> Thoughts @kfswain @nirrozenbaum @ahg-g

> agreed. we should probably respect both configurations with some precedence plus add deprecation message to both documentation (in code) and log (only if the env var is used).
the precedence should be - read config from feature gate if exist (if not fallback to default).
if env var exist - override feature gate (plus deprecation log message).

I have rebased the code, and implemented the above suggested deprecation logic.

The trouble shooting guide and the release notes have been updated to reflect this.

kfswain commented:
/approve

Will leave the stamp to someone else

nirrozenbaum commented on pkg/epp/config/config.go:
```diff
@@ -16,9 +16,14 @@ limitations under the License.
-	SchedulerConfig *scheduling.SchedulerConfig
+	SchedulerConfig          *scheduling.SchedulerConfig
+	FeatureConfig            map[string]bool
+	SaturationDetectorConfig saturationdetector.Config
```
why not pointer?

shmuelk commented on cmd/epp/runner/runner.go:
```diff
@@ -415,25 +417,69 @@ func (r *Runner) parsePluginsConfiguration(ctx context.Context, ds datastore.Dat
+	if _, ok := os.LookupEnv(enableExperimentalDatalayerV2); ok {
+		logger.Info("Enabling the experimental Data Layer V2 using environment variables is deprecated")
+		cfg.FeatureConfig[datalayer.FeatureGate] = env.GetEnvBool(enableExperimentalDatalayerV2, false, logger)
+	}
```
The logging done by the utility when the default value is used will not tell the user that use of the variable is deprecated.
